rfcs/SubschemaComposability.md

```diff
@@ -0,0 +1,74 @@
+
+The below RFC is meant to be useful across all varieties of schema composition. Because all forms of schema composition require a composite schema, we can speak of "composability" without referring to the timing of composition or the type of execution. However, when relevant, reference will be made to underlying technical factors which affect "composability." Practically speaking, the method of composite schema execution matters more than the timing of composition. If a request against the composite schema must ultimately be translated to request(s) against the still extant source subschemas, then the composition must be -- in some sense -- reversed during execution, an additional strenuous requirement.
+
+# Scalars
+
+1. The specified scalars (Int, Float, String, Boolean, ID) can each be merged across subschemas, because they are identical in all subschemas. 
+2. Custom scalars can be merged across subschemas, _as long as the parsing/serialization of values is performed in the identical manner across all subschemas._
+   - This important condition/qualification may be difficult to verify!
+   - The `specifiedByURL` field may be helpful in this regard.
+
+# Enums
+
+1. Overlapping enum types where the types define identical sets of values can always be merged, as they are identical in all subschemas.
+2. Subschemas with overlapping enum types where the disparate types define different value sets are only sometimes composable.
+   - When using `Single-service composite schema execution`, the enum types can be merged as a union of values from all subschemas.
+   - When using `Subschemas as remote GraphQL services`
+     - If the enum types are used only in output types, the enum types can be merged as a union of values from all subschemas.
+     - If the enum types are used in input types, the enum types can be merged if and only if there is a way to translate, for any given subschema, every enum value that is defined only in the composite schema to an enum value defined for that subschema.
+
+# Input Object Types
+
+1. Overlapping input object types can be merged, as long as the overlapping input fields can be merged.
+2. If some of the types define fields not defined by the other types, the fields of these input objects may no longer affect the execution of portions of the subschema, in ways that may be unpredictable for users of the composite schema. Although clear documentation may ameliorate this issue, overall, the user experience may suffer. 
+
+# Input Object Fields
+
+1. Overlapping input object fields can be merged as long as the fields define the same type, allowing for variations in nullability.
+2. If any of the field types contain non-nullable portions, the field types must be merged as non-nullable for that portion.
+3. TODO: discuss default values.
```

Comment:
`3. ... default values` is interesting. Is a default value normally filled in by an executor? I suppose this could vary across implementations.

Reply:
Default values are encoded by subschemas and substituted by the executor if no value is sent. If there is a discrepancy between subschemas with regards to the default, ie whether one is defined or what the value is, I think a merge should still succeed, and I basically agree that we don't really need to specify whether/how the composite schema should define a default. :)

Reply:
I do not have direct experience with Nautilus but from the docs it looks like types are composed if they have the same name, and that cross service runtime resolution is by using the Node interface and node root query field.

It does not say what happens if object types of the same name do not implement Node or if the node root query field is absent. Presumably, composition will fail or the types must be namespaced.

Assuming the latter, the use of the Node interface does not seem to be a separate way to maintain composability per se; rather, the composition proceeds by name spacing by default unless type composability is possible. This is certainly a sane default rather than simply failing.

I am making a lot of assumptions from the docs I could find, please let me know how/if I have misunderstood how Nautilus works. :)